README.md
# ecs-stale-ec2-rescaler

[![Tags](https://img.shields.io/github/tag/civitaspo/ecs-stale-ec2-rescaler.svg?style=flat-square)](https://github.com/civitaspo/ecs-stale-ec2-rescaler/tags)
[![Docker Pulls](https://img.shields.io/docker/pulls/civitaspo/ecs-stale-ec2-rescaler.svg?style=flat-square)](https://hub.docker.com/r/civitaspo/ecs-stale-ec2-rescaler)

Docker Image for ECS to terminate a stale container instance that belongs to a autoscaling group.

# How this works

1. Polling the latest `/var/log/ecs/ecs-agent.log` until the below error message is detected.
    - `Error getting message from ws backend`
    - `Error response from daemon: conflict: unable to delete`
    - `Duplicate ENI attachment message`
1. Notify the detection to Slack if `SLACK_URL` is specified.
1. Execute `aws ecs put-attributes --cluster $ECS_CLUSTER --attributes $ATTRIBUTES` if `ATTRIBUTES_FOR_STALE_EC2` is specified.
1. Execute `aws autoscaling terminate-instance-in-auto-scaling-group --instance-id $INSTANCE_ID --no-should-decrement-desired-capacity` if `TERMINATE_STALE_EC2=true`.
1. Execute `aws autoscaling detach-instances --instance-id $INSTANCE_ID --auto-scaling-group-name $AUTOSCALING_GROUP_NAME --no-should-decrement-desired-capacity` if `ENTER_STANDBY_STALE_EC2=true` and `TERMINATE_STALE_EC2=false`.
1. Execute `aws autoscaling detach-instances --instance-id $INSTANCE_ID --auto-scaling-group-name $AUTOSCALING_GROUP_NAME --no-should-decrement-desired-capacity` if `DETACH_STALE_EC2=true` and `TERMINATE_STALE_EC2=false`. 

# Environment Variables

- `POLLING_MAX_ATTEMPTS`: Max number of attempts for polling. (default: `60`)
- `POLLING_INTERVAL`: Interval seconds for polling. (default: `1`)
- `DUPLICATE_ENI_ATTACHMENT_PER_HOUR_THRESHOLD`: Threshold for number of 'Duplicate ENI attachment message' per hour. (default: `50`)
- `TERMINATE_STALE_EC2`: Terminate stale ec2 or not. (default: `true`)
- `ENTER_STANDBY_STALE_EC2`: Enter standby stale ec2 or not. Ignore this option `TERMINATE_STALE_EC2=true` (default: `false`)
- `DETACH_STALE_EC2`: Detach stale ec2 or not. Ignore this option `TERMINATE_STALE_EC2=true` (default: `false`) 
- `ATTRIBUTES_FOR_STALE_EC2`: Attributes for stale ec2. This must be 'key=value' strings delimited by a space. (default: empty string)
- `INSTANCE_IDENTITY_URL`: The URL to get the instance identity. (default: `"http://169.254.169.254/latest/dynamic/instance-identity/document"`)
- `SLACK_URL`: Slack webhook url for the notification. (optional)
- `SLACK_ADDITIONAL_MESSAGE`: Slack addtional message for the notification. (optional)
- `SLACK_CHANNEL`: Slack channel for the notification. (optional)
- `SLACK_ICON_EMOJI`: Slack icon emoji for the notification. (optional)

# Issues that this repository concerns

- [amazon-ecs-agent#1698 Docker corrupted](https://github.com/aws/amazon-ecs-agent/issues/1698)
- [amazon-ecs-agent#1667 One task prevents from all of the other in instance to change from PENDING to RUNNING](https://github.com/aws/amazon-ecs-agent/issues/1667)
- [amazon-ecs-agent#1980 Cleanup is not working when ECS mamanged image is running in non-managed container](https://github.com/aws/amazon-ecs-agent/issues/1980)

# IAM Privileges
- (required) `ec2:DescribeTags`
- (optional) `ecs:ListContainerInstances`
- (optional) `ecs:PutAttributes`
- (optional) `autoscaling:TerminateInstanceInAutoScalingGroup`
- (optional) `autoscaling:EnterStandby`
- (optional) `autoscaling:DetachInstances`

# Example

- [Example Task](./example/ecs-task-cli-input.json)
- [Example Task Role](./example/ecs-task-role.json)
- [Example Service](./example/ecs-service-cli-input.json)

# ChangeLog

[CHANGELOG.md](./CHANGELOG.md)

# License

[Apache License 2.0](./LICENSE.txt)

# Author

@civitaspo



CHANGELOG.md
0.0.4 (2019-06-23)
==================

* Support ecs-agent 1.28.0 error message about `Duplicate task-eni attachment message`. (#9)

0.0.3 (2019-04-15)
==================

* Detect the 'Error response from daemon: conflict: unable to delete' error. (#5)

0.0.2 (2019-02-24)
==================

* Detect duplicate eni attachment message abnormality (#3)

0.0.1 (2018-12-31)
==================

* First Release


entrypoint.sh
#!/usr/bin/env bash

set -eu
set -o pipefail

# Define loggers
function now() {
    date +'%Y-%m-%d %H:%M:%S %z'
}
function __log() {
    local -u -r level="${1:-unknown}"
    local -r message="${2:-}"
    echo "$(now) [$level] $message"
}
function __info_log() {
    __log info "${1:-}"
}
function __warn_log() {
    __log warn "${1:-}"
}
function __error_log() {
    __log error "${1:-}" 1>&2
}

__info_log "This script polls ecs-agent logs to detect the ecs container instance stale state. When the state is detected, this script terminates the instance."
__info_log "See https://github.com/civitaspo/ecs-stale-ec2-rescaler for more information."

# Declare variables
declare -r  APP_NAME=ecs-stale-ec2-rescaler
declare -ir POLLING_MAX_ATTEMPTS=${POLLING_MAX_ATTEMPTS:-60}
declare -ir POLLING_INTERVAL=${POLLING_INTERVAL:-1}
declare -ir DUPLICATE_ENI_ATTACHMENT_PER_HOUR_THRESHOLD=${DUPLICATE_ENI_ATTACHMENT_PER_HOUR_THRESHOLD:-50}

declare -r INSTANCE_IDENTITY_URL=${INSTANCE_IDENTITY_URL:-http://169.254.169.254/latest/dynamic/instance-identity/document}
declare -i num_attempts=0
while [ -z "$(curl -s $INSTANCE_IDENTITY_URL | jq -r .instanceId)" ]; do
    if ((num_attempts == POLLING_MAX_ATTEMPTS)); then
        __error_log "Max attpempts(=$POLLING_MAX_ATTEMPTS) is exceeded because $INSTANCE_IDENTITY_URL did not become available."
        exit 1
    fi
    __info_log "Wait until $INSTANCE_IDENTITY_URL become available. ($((num_attempts++))s)"
    sleep $POLLING_INTERVAL
done
unset num_attempts

declare -r INSTANCE_ID=$(curl -s ${INSTANCE_IDENTITY_URL} | jq -r .instanceId)
declare -r REGION=$(curl -s $INSTANCE_IDENTITY_URL | jq -r .region)
# NOTE: Detect the autoscaling group just for logging, so do not exit if the detection is failed.
declare -r AUTOSCALING_GROUP_NAME=$(
    aws ec2 describe-tags \
        --region $REGION \
        --filters "Name=resource-id,Values=$INSTANCE_ID" \
                  "Name=key,Values=aws:autoscaling:groupName" \
        | jq -r '.Tags[0].Value' \
        || echo "AutoDetectionFailed"
    )

declare -r SLACK_URL=${SLACK_URL:-}
declare -r SLACK_ADDITIONAL_MESSAGE=${SLACK_ADDITIONAL_MESSAGE:-}
declare -r SLACK_CHANNEL=${SLACK_CHANNEL:-}
declare -r SLACK_ICON_EMOJI=${SLACK_ICON_EMOJI:-}

__info_log "Polling until errors are catched."
declare STALE_STATE_CAUSE=""
while true; do
    if ls /var/log/ecs/ecs-agent.log* | sort -n | tail -n1 | xargs -I{} grep -r "Error getting message from ws backend" {} >/dev/null; then
        STALE_STATE_CAUSE="'Error getting message from ws backend' is occurred"
        break
    fi
    if ls /var/log/ecs/ecs-agent.log* | sort -n | tail -n1 | xargs -I{} grep -r "Error response from daemon: conflict: unable to delete" {} >/dev/null; then
        STALE_STATE_CAUSE="'Error response from daemon: conflict: unable to delete' is occurred"
        break
    fi
    for n in $(grep -P 'Duplicate (ENI|task-eni) attachment message' /var/log/ecs/ecs-agent.log* | cut -f2 -d: | sort -n | uniq -c |  xargs -n2 echo | cut -f1 -d' '); do
        if ((n > DUPLICATE_ENI_ATTACHMENT_PER_HOUR_THRESHOLD)); then
            STALE_STATE_CAUSE="'Duplicate ENI attachment message' count exceeds $DUPLICATE_ENI_ATTACHMENT_PER_HOUR_THRESHOLD/h"
            break 2
        fi
    done
    sleep $POLLING_INTERVAL
done

declare -r MESSAGE="Detect stale state:[$STALE_STATE_CAUSE], so terminate $INSTANCE_ID in asg:$AUTOSCALING_GROUP_NAME."
declare -r SLACK_MESSAGE=":hammer_and_wrench: $MESSAGE $SLACK_ADDITIONAL_MESSAGE"
__warn_log "$MESSAGE"

if [ ! -z "$SLACK_URL" ]; then
    curl -X POST \
         -H 'Content-type: application/json' \
         --data "
             {
                 \"text\": \"$SLACK_MESSAGE\",
                 \"channel\":\"$SLACK_CHANNEL\",
                 \"icon_emoji\":\"$SLACK_ICON_EMOJI\",
                 \"username\":\"$APP_NAME\"
             }
             " \
         $SLACK_URL
fi

# Terminate the container instance.
aws autoscaling terminate-instance-in-auto-scaling-group \
    --instance-id $INSTANCE_ID \
    --no-should-decrement-desired-capacity \
    --region $REGION

# Sleep for 200 seconds to prevent this script from looping.
# The instance should be terminated by the end of the sleep.
sleep 200
